src/all/mangadex/src/eu/kanade/tachiyomi/extension/all/mangadex/MangaDexHelper.kt

```diff
@@ -283,14 +283,26 @@ class MangaDexHelper(lang: String) {
         coverFileName: String?,
         coverSuffix: String?,
         lang: String,
+        preferSourceLangTitle: Boolean,
     ): SManga = SManga.create().apply {
         url = "/manga/${mangaDataDto.id}"
         val titleMap = mangaDataDto.attributes!!.title
         val dirtyTitle =
-            titleMap.values.firstOrNull() // use literally anything from title as first resort
-                ?: mangaDataDto.attributes.altTitles
-                    .find { (it[lang] ?: it["en"]) !== null }
-                    ?.values?.singleOrNull() // find something else from alt titles
+            titleMap[lang] ?: if (preferSourceLangTitle) {
+                mangaDataDto.attributes.altTitles
+                    .find { it[lang] !== null } // try use the source language
+                    ?.values?.singleOrNull()
+                    ?: titleMap.values.firstOrNull() // fallback to the original language title
+                    ?: mangaDataDto.attributes.altTitles
+                        .find { it["en"] !== null } // if all else fails, try to find an english title
+                        ?.values?.singleOrNull()
+            } else {
+                titleMap.values.firstOrNull() // use literally anything from title as first resort
+                    ?: mangaDataDto.attributes.altTitles
+                        .find { (it[lang] ?: it["en"]) !== null }
+                        ?.values?.singleOrNull() // find something else from alt titles
+            }
```

Comment:
Consider:
```kt
private fun List<Map<String, String>>.findTitleByLang(lang: String): String? =
    firstOrNull { it[lang] != null }?.values?.singleOrNull()
```

```kt
val dirtyTitle = with(mangaDataDto.attributes) {
    titleMap[lang] ?: altTitles.run {
        val mainTitle = titleMap.values.firstOrNull()
        val langTitle = findTitleByLang(lang)
        val enTitle = findTitleByLang("en")

        if (preferSourceLangTitle) {
            listOf(langTitle, mainTitle, enTitle).firstNotNullOfOrNull { it }
        } else {
            listOf(mainTitle, langTitle, enTitle).firstNotNullOfOrNull { it }
        }
    }
}
```
    
or

```kt
val altTitles = mangaDataDto.attributes.altTitles
val dirtyTitle = titleMap[lang] ?: if (preferSourceLangTitle) {
    listOf(
        altTitles.findTitleByLang(lang),
        titleMap.values.firstOrNull(),
        altTitles.findTitleByLang("en"),
    ).firstNotNullOfOrNull { it }
} else {
    listOf(
        titleMap.values.firstOrNull(),
        altTitles.findTitleByLang(lang),
        altTitles.findTitleByLang("en"),
    ).firstNotNullOfOrNull { it }
}
```

(needs testing)

Reply:
The only time it will fall back to using English is if `altTitles` doesn't have the chosen extension language and `titles` has nothing. This is an existing assumption(?), and if both those are missing it is then assuming English will be available in `altTitles`.

Rather than removing it, I think adding another fallback would be preferable. So, for example:
```kt
titleMap.values.firstOrNull(),
altTitles.findTitleByLang(lang),
altTitles.findTitleByLang("en"),
altTitles.firstOrNull(),
```

This will only fail if both `title` and `altTitles` is missing, but at that point, there must be something seriously wrong.

You can check the old history here: https://github.com/funkyhippo/tachiyomi-extensions/blame/1c8478fcca73ce1c8d36bb6be021253b4be1ccbe/src/all/mangadex/src/eu/kanade/tachiyomi/extension/all/mangadex/MangaDexHelper.kt